src/gasman.c

```diff
@@ -2081,54 +2081,65 @@ void CheckMasterPointers( void )
 }
 
 
-/****************************************************************************
-**
-*F  SwapMasterPoint( <bag1>, <bag2> ) . . . swap pointer of <bag1> and <bag2>
-*/
-void SwapMasterPoint (
-    Bag                 bag1,
-    Bag                 bag2 )
+// Swap the master pointers of bag1 and bag2
+//
+// We need to make sure the correct bags are GCed. There are various things to
+// consider.
+//
+// As changed bags are stored as a singly linked list, we can't easily take
+// anything out of the list of changed bags. Also, as the changed bags are stored
+// in a linked list with the pointer stored inside the bag itself, when we swap
+// PTR_BAG, we swap LINK_BAG as well.
+//
+// If both bags are old, if the original bag1 was changed we need to now mark
+// bag2 changed (and vice-versa).
+//
+// If both bags are young, GAP assumes they are both changed.
+//
+// If bag1 was young and bag2 was old, then after swapping, bag1 is old and bag2 is
+// young (the 'young'ness moves with the contents)
+//
+// Mark bag1 changed if bag2 was previously changed (for same reason as above).
+//
+// More importantly, bag2 is now young, but the only reference to bag2 might be
+// in an old, unchanged bag, so it would get GCed.
+//
+// We therefore force bag2 onto the ChangedBags list (we can't use CHANGED_BAG,
+// as it skips young bags). Gasman treats anything on the ChangedBags list as a
+// root, not garbage collecting it or anything it references.
```

Comment:
Actually that is not quite correct: GASMAN treats every old bag on the changed list as a root, but young bags get special treatment: They are not treated as roots, but nevertheless marked and thus not collected.

"Gasman never garbage collects any bags on the list of changed bag"